include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +96,182 @@ parse(
+   should be defined to consume the input from [`it`, `end`) into `arg_i`.<br>
 
-   This `parse` function is defined for `char`s so that it matches the BNF
-   string literals in [@ref it, @ref end).
-
-   The function returns `true` and resets @ref ec when the input is
+   The function returns `true` and resets `ec` when the input is
    parsed successfully. If the input cannot be parsed into the BNF elements,
-   the function returns `false` and sets the error number in @ref ec to the
+   the function returns `false` and sets the error number in `ec` to the
    appropriate error number defined in @ref boost::urls::error.
 
    @par Example
    @code
-   #include <iostream>
-   #include <string>
-
-   #include <boost/url/bnf/char_set.hpp>
-   #include <boost/url/bnf/parse.hpp>
-   #include <boost/url/error.hpp>
-
-   namespace boost {
-   namespace urls {
-   namespace detail {
-   // BNF element to parse a non-empty sequence of alpha chars ([a-zA-Z]+)
-   struct alpha_string_bnf {
-     // reference to a std::string where we should store the results
-     std::string &v;
-
-     // parsing an alpha_string_bnf element
-     friend bool parse(char const *&it, char const *end, error_code &ec,
-                       alpha_string_bnf &&t) {
-       // is_alpha functor
-       bnf::alpha_chars is_alpha;
-
-       // check if empty
-       if (it == end) {
-         ec = error::incomplete;
-         return false;
-       }
-
-       // check if not alpha at all
-       if (!is_alpha(*it)) {
-         ec = error::bad_alpha;
-         return false;
-       }
-       t.v.push_back(*it);
-       ++it;
-
-       // consume more alphas
-       while (it != end && is_alpha(*it)) {
-         t.v.push_back(*it);
-         ++it;
-       }
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (bnf::parse(it, end, ec, detail::alpha_string_bnf{alpha_l}, '.',
-                    detail::alpha_string_bnf{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
+   if (bnf::parse(it, end, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
+     std::cout << "Range [it, end) parsed successfully\n";
    }
    @endcode
 
    @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
 
-   @see https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+     Internal calls to `parse` overloads for `args` may throw, although they should
+     prefer setting `ec` instead.
```

Comment:
Stock phrase: "Basic guarantee"

Reply:
I'm not sure I understand how to replicate this one. The only place I can find the term `Basic guarantee` is in `url/url.hpp` so I can't infer the best approach:

```
@par Exception Safety
Basic guarantee.
Calls to allocate may throw.
```

Should I replace the current description or just add the `Basic guarantee` prefix?

Should the `Exception Safety` paragraph only describe one of the four [Exception safety guarantees](https://en.cppreference.com/w/cpp/language/exceptions) instead of discussing it?

1. Nothrow (or nofail) exception guarantee -- the function never throws exceptions. Nothrow (errors are reported by other means or concealed) is expected of destructors and other functions that may be called during stack unwinding. The destructors are noexcept by default. (since C++11) Nofail (the function always succeeds) is expected of swaps, move constructors, and other functions used by those that provide strong exception guarantee.
2. Strong exception guarantee -- If the function throws an exception, the state of the program is rolled back to the state just before the function call. (for example, std::vector::push_back)
3. Basic exception guarantee -- If the function throws an exception, the program is in a valid state. No resources are leaked, and all objects' invariants are intact.
4. No exception guarantee -- If the function throws an exception, the program may not be in a valid state: resource leaks, memory corruption, or other invariant-destroying errors may have occurred.

Reply:
This is a good stock phrase (if the function allocates)

```
@par Exception Safety
Basic guarantee.
Calls to allocate may throw.
```

In this case we don't know if a BNF element will allocate since the function is generic. And we don't know if a BNF element will throw. So we might use this instead:

```
@par Exception Safety
Defined by the types of `args`.
```

or something like that.

Reply:
OK. Changing to `defined by the types of args` then. 

So if `Defined by the types of args.` is correct, I wasn't wrong in assuming this section is for discussing exception safety in general (like cpp reference does) instead of just choosing one of the categories, right? And there's no stock phrase in that case because it's not the same logic, right?

Reply:
Oh... by the way... I think it's possible to officially make `parse` `noexcept` depending on the types of `args` with some little metaprogramming spaghetti.

Reply:
>  I think it's possible to officially make parse noexcept depending on the types of args with some little metaprogramming spaghetti.

Maybe, but I prefer to avoid such complexity unless there is *overwhelming evidence* of significant benefit, which I am not seeing here. Simpler is usually better.